### DISCUSSION TITLE
I Can't Add New Timelines To Ghosts Through API

### DISCUSSION THREAD

**Discussion by danross826** [cmu-sei/GHOSTS#271] [Q&A] [answered]
A coworker of mine set up the Ghosts API while I've been working on the front end of our app. I can add new Machines through the API post request as well as retrieve them with a get request, but I can't add new timelines through Ghosts timeline post request. I don't get an error message, and it seems to work when I send the request. However, when I use the get request to pull down all the timelines in a machine, I just get an empty object. Nothing appears to be added despite the lack of an error message. Can anyone help me with this?

**Comment by sei-dupdyke** [comment]
Hi, thanks for checking out the project.

Hmm, it might be a couple of things — let's start with what you're posting as a timeline. Would you please post the json payload for POST to `/api/timelines`?

**Reply by danross826** [reply]
This is the json load:

`{
    "machineId": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "type": 0,
    "activeUtc": "2024-02-08T21:25:31.419Z",
    "status": 0,
    "update": {
        "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
        "status": 0,
    "TimeLineHandlers": [
        {
            "HandlerType": "NpcSystem",
            "Initial": "",
            "UtcTimeOn": "00:00:00",
            "UtcTimeOff": "24:00:00",
            "Loop": false,
            "TimeLineEvents": [
                {
                    "Command": "Stop",
                    "CommandArgs": [],
                    "DelayAfter": 0,
                    "DelayBefore": 0
                }
            ]
        },
       {
            "HandlerType": "NpcSystem",
            "Initial": "",
            "UtcTimeOn": "00:00:00",
            "UtcTimeOff": "24:00:00",
            "Loop": false,
            "TimeLineEvents": [
                {
                    "Command": "Start",
                    "CommandArgs": [],
                    "DelayAfter": 0,
                    "DelayBefore": 0
                }
            ]
        }
    ]
    }
}`

**Reply by sei-dupdyke** [reply]
That's good json. Do you see anything in the API container logs? Do something like `docker logs ghosts-api` after a post and see if there are any errors listed?

**Reply by danross826** [reply]
When I try to use that command to view the logs, I'm getting a postgres related error I believe.

![image](https://github.com/cmu-sei/GHOSTS/assets/20523362/7f2688a7-5a54-4626-b2e0-0ffd67132fbf)

**Reply by sei-dupdyke** [reply]
When you do a `docker ps --all` are all the containers running?

**Reply by danross826** [reply]
I believe so.

![image](https://github.com/cmu-sei/GHOSTS/assets/20523362/cb3c5a32-9bc6-49de-9b5f-c44a36f96c62)

**Reply by sei-dupdyke** [reply]
Hmm, you should have three containers like this:
![image](https://github.com/cmu-sei/GHOSTS/assets/43444464/7664a3fa-d3bc-4f88-9d9a-d3efb398fa67)

**Reply by danross826** [reply]
Okay, I got all three containers up and running but it's still throwing me the same error when I try accessing the logs.

![image](https://github.com/cmu-sei/GHOSTS/assets/20523362/f4b4c845-d65c-4d30-9501-7594385fd7d9)

**Reply by sei-dupdyke** [reply]
If you used the docker-compose file to stand up the containers (I assume you did), you may need to chown [the mounted database files set by this part of that compose file](https://github.com/cmu-sei/GHOSTS/blob/master/src/Ghosts.Api/docker-compose.yml#L17C6-L17C43) ` - ./db_data:/var/lib/postgresql/data`.

```bash
sudo chown -R 777 db_data
```

or whatever account the docker daemon is running under (in ubuntu, you typically need to sudo or su to get to docker, its not installed with regular user permissions).

**Reply by danross826** [reply]
I got access to the logs and cleared them. I then tried to do my posts again. None of them are showing up in the logs. The only time something shows up in the logs is when I purposely enter bad json then I get an error and that shows up. However, the original reason I started this is because I'm putting in good json with the timeline and not getting any errrors at all.

**Reply by sei-dupdyke** [reply]
Hmm, I tested here locally and this works. If you are looking in the database, you should see a record in machine_updates with the status of 0. Once it is processed that status is changed to -9. If you aren't seeing a record in that table, then we have a different problem with that timeline being saved.

**Reply by danross826** [reply]
I got access to the machine_updates table in the db. They're all status 0. However, I'm noticing that it doesn't have records after February 09 even though I have made post requests since then.

**Reply by danross826** [reply]
Okay, I was wrong. The changes are appearing in machine_updates, but they're all status 0.

**Reply by danross826** [reply]
I also seem to now be unable to get my client id via the get request address listed for ghosts-api.

![image](https://github.com/cmu-sei/GHOSTS/assets/20523362/d3f843a2-999d-412f-b8cf-ca79a8d80bae)

**Reply by sei-dupdyke** [reply]
Hi sorry I was traveling and just catching up. Is this from the client's log or are you trying to hit a client_x endpoint directly? (You can't hit any endpoint starting with "client" as the actual client sends information in the webrequest headers that you'd have to emulate)

**Reply by danross826** [reply]
It's from the /api/clientid. The get requests that post and show machines work, but this one does not.

![image](https://github.com/cmu-sei/GHOSTS/assets/20523362/b8e7a782-34f3-41a8-bc49-a6f45de69ab0)

**Reply by sei-dupdyke** [reply]
Yep, that is for clients — they check in and get an GUID Id from the system (if one does not exist in the instance folder). For us humans, use the Machine endpoints to get the Id of some client. 

GET /api/machines/list — is a fast way to find a machine if you know its name.

**Reply by danross826** [reply]
I've been reading the code, and I'm trying to get a handle on how ghosts works with regard to adding new timelines to a machine. Correct me if I'm wrong, but when you post a timeline to a machine, it goes into machine updates and will be added to the machine the next time the client connects to it and it processes the machine updates, right?

**Reply by sei-dupdyke** [reply]
That is correct, but in v8 clients will initiate a socket connection with the server and so the "check-in" time should be very fast :)

**Reply by danross826** [reply]
I don't know if I have incomplete documentation, but is there something that shows the workflow of an average ghosts client that gets its info from an api server where you can add new timelines with actions?

**Reply by sei-dupdyke** [reply]
No, that doesn't currently exist, sorry. I'm way behind on docs.

**Reply by danross826** [reply]
I don't suppose you have an updated documentation regarding the application.json file that the client needs to access the API because the example in the docs looks different than the actual file that comes with the client? 

![image](https://github.com/cmu-sei/GHOSTS/assets/20523362/dd9f6342-db10-4fee-876d-760b545deb2f)

![image](https://github.com/cmu-sei/GHOSTS/assets/20523362/8da7a59f-af27-4403-ad6b-748afb12a56e)

**Reply by sei-dupdyke** [reply]
No I don't, but I suspect you can find the ones that you're looking for in the docs — they've largely just moved, not been radically changed.

**Reply by danross826** [reply]
My ghosts client doesn't seem to be accessing the api despite the fact that I know my api url in the example above is correct because when I put it into the browser navbar, the api comes up. Whenever I activate the client, it will follow its local timeline file but not contact the api. If I remove the local timeline file then it does nothing at all.

**Reply by sei-dupdyke** [reply]
Hmm

> My ghosts client doesn't seem to be accessing the api despite the fact that I know my api url in the example above is correct because when I put it into the browser navbar, the api comes up. 

Would you please post up your client config/application.json file here?

> Whenever I activate the client, it will follow its local timeline file but not contact the api. 

This is expected behavior. Local timeline is default until overwritten by API — so if you can't get to API, it will just continue to execute local timeline indefinitely.

> If I remove the local timeline file then it does nothing at all.

This is also expected behavior. A default timeline is required. It can have no activities in it, but it must exist.

**Reply by danross826** [reply]
I'm posting my application.json.
[application.json](https://github.com/cmu-sei/GHOSTS/files/14516789/application.json)

**Reply by sei-dupdyke** [reply]
Hmm, that seems legitimate. Confirm you can hit http://10.128.241.68:5000/ in your browser on the client machine?

**Reply by danross826** [reply]
Yes.

![image](https://github.com/cmu-sei/GHOSTS/assets/20523362/154286b3-f338-4bde-b0a7-d1ea35c07bb4)

**Reply by danross826** [reply]
I used wireshark to analyze the traffic between the vm with the client and the api server. It seems like they're communicating, but when the api server communicates back with the client, it gets a 404 error. I don't know if that means anything to you. Like, I'm seeing the api server trying to communicate back to the client with the client updates api address, and it throws a 404.

**Reply by sei-dupdyke** [reply]
Note: If you are trying to use those client endpoints they will fail, the client uses some custom headers to identify themselves.

404 is expected for some endpoints - like if there is no client update currently, then a 404 is expected behavior.

**Reply by danross826** [reply]
Well, I'm not sure if I'm misunderstanding something with order of operations, but I've submitted timelines for this machine with the post api addresses and they just stay in machine updates with a status of 0. I thought when the client communicated with the api server that they'd be processed and my timelines would show up in the timelines table. I don't know if the server communicating back with the client and getting a 404 might be related to that not working how it's supposed to.

**Reply by danross826** [reply]
I think it's possible that the return messaging that's getting the 404 error is hitting a port that the client isn't listening to. Wireshark says it's hitting port 52550 while when I check my windows machine with the client, it says it's listening on port 135. I'm looking to see if there's a way to change the port on either the client or the api server. I didn't set up the api server so I'm just going through the files to see if there's something that might correspond to this.

**Comment by sei-dupdyke** [comment]
> Well, I'm not sure if I'm misunderstanding something with order of operations, but I've submitted timelines for this machine with the post api addresses and they just stay in machine updates with a status of 0. I thought when the client communicated with the api server that they'd be processed and my timelines would show up in the timelines table. I don't know if the server communicating back with the client and getting a 404 might be related to that not working how it's supposed to.

Hmm, I think you have it right:

- client checks in via web request with information about itself — including its globally unique MachineId, but there are no updates for that MachineId, so client receives 404 status code from the server. Client understands this as "no updates found for me".
- you submit a machine timeline for a particular MachineId via the API.
- the server holds it in machine_updates table until that client checks in (basically asking, "do you have anything new for me to do?)"
- client checks in again, that update you posted is found by matching the MachineId of the incoming client request to the machine you indicated that the update was for, and the response code is now 200, which tells the client "here is a new thing for you to do".
- server marks that update as processed ("-9" IIRC).

> I think it's possible that the return messaging that's getting the 404 error is hitting a port that the client isn't listening to. Wireshark says it's hitting port 52550 while when I check my windows machine with the client, it says it's listening on port 135. I'm looking to see if there's a way to change the port on either the client or the api server. I didn't set up the api server so I'm just going through the files to see if there's something that might correspond to this.

What you are seeing in wireshark is that the client makes a request to the server on 52550 and the latter returns 404. The client receives this on an ephemeral port (135 in this case), like this:

client --> server:52550 -> responds with "404" --> received on client:135.

---

couple of things to check:

- what is the uuid in ./instance/id.json? (this is the global Id for this client)
- if you do a `select * From machine_updates;` does the machineId match the above?

**Reply by danross826** [reply]
Okay, so you guys recently updated ghosts so I reinstalled docker and the new version of ghosts api and all that. I also downloaded a new client on my other machine that I'm trying to run ghosts on. Okay, so I get that when you activate a new client it creates an id.json file that has that client's new machine id, and it adds that id to the machines table in the ghosts api server. So, I should be able to create new timelines with that machine id and it should be able to run and follow those timelines. I've attempted to do this on swagger with the example timeline you have on the timeline post api address, and now I'm just getting a 500 error every time.

![image](https://github.com/cmu-sei/GHOSTS/assets/20523362/247984a9-1cad-4d2a-85f0-ddf2c6082990)

**Comment by sei-dupdyke** [comment, marked as answer]
My apologies, I finally realized this morning what the issue is that you are pointing out!

For `POST /api/machineupdates` the previously generated sample json payload was:

```json
{
  "machineId": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "type": "Timeline",
  "activeUtc": "2024-03-23T13:39:44.558Z",
  "status": "Active",
  "update": {
    "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "status": "Run",
    "timeLineHandlers": [
      {
        "handlerType": "BrowserFirefox",
        "initial": "string",
        "utcTimeOn": {
          "ticks": 0
        },
        "utcTimeOff": {
          "ticks": 0
        },
        "utcTimeBlocks": [
          {
            "ticks": 0
          }
        ],
        "handlerArgs": {
          "additionalProp1": "string",
          "additionalProp2": "string",
          "additionalProp3": "string"
        },
        "loop": true,
        "timeLineEvents": [
          {
            "trackableId": "string",
            "command": "string",
            "commandArgs": [
              "string"
            ],
            "delayAfter": "string",
            "delayBefore": "string"
          }
        ],
        "scheduleType": "Other",
        "schedule": "string"
      }
    ]
  }
}
```

Any of the settings with "ticks" are timespans types, and they don't serialize properly for our purposes here in this method. I fixed this by explicitly creating a sample object that gets displayed now with version 8.0.7 of the container:

```json
{
  "machineId": "8da92fc7-01af-4261-80e0-33539a7f251a",
  "type": "TimelinePartial",
  "activeUtc": "2024-03-23T14:14:37.411893Z",
  "status": "Active",
  "update": {
    "id": "102d599b-df89-4ed5-891f-e0a09f14210e",
    "status": "Run",
    "timeLineHandlers": [
      {
        "handlerType": "BrowserFirefox",
        "initial": "https://cmu.edu",
        "utcTimeOn": "00:00:00",
        "utcTimeOff": "23:59:59",
        "utcTimeBlocks": null,
        "handlerArgs": {},
        "loop": true,
        "timeLineEvents": [
          {
            "trackableId": null,
            "command": "browse",
            "commandArgs": [
              "https://sei.cmu.edu"
            ],
            "delayAfter": 30000,
            "delayBefore": 0
          }
        ],
        "scheduleType": "Other",
        "schedule": null
      }
    ]
  }
}
```

Sorry for the confusion, I will try to be sure to check these swagger functions more closely with future updates.
